nucleus/v1/types.go

```diff
@@ -0,0 +1,187 @@
+package v1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+
+// NucleusHub represents a deployment of nucleus hub component.
+type NucleusHub struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec represents a desired deployment configuration of nucleus hub
+	Spec NucleusHubSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status represents the current status of nucleus hub
+	// +optional
+	Status NucleusHubStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// NucleusHubSpec represents a desired deployment configuration of nucleus hub.
+type NucleusHubSpec struct {
+	// Components represents the components to be installed for nucleus hub.
+	// +required
+	Components NucleusHubComponents `json:"components" protobuf:"bytes,1,opt,name=components"`
+}
+
+// NucleusHubStatus represents the current status of nucleus hub.
+type NucleusHubStatus struct {
+	// Conditions contain the different condition statuses for this work.
+	// Valid condition types are:
+	// Deployed: components in hub is deployed
+	// Ready: components in hub are ready to serve
+	// Degraded: components in hub does not match the desired configuration and only provide
+	// degraded service.
+	Conditions []StatusCondition `json:"conditions" protobuf:"bytes,1,opt,name=conditions"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// NucleusHubList is a collection of nucleus hub.
+type NucleusHubList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is a list of nucleus hub.
+	Items []NucleusHub `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// NucleusHubComponents represents the components to be installed for nucleus hub.
+type NucleusHubComponents struct {
+	// Registration represents the desired configuration of registration controller installed on hub.
+	// +required
+	Registration RegistrationComponent `json:"registration" protobuf:"bytes,1,opt,name=registration"`
+}
+
+// RegistrationComponent represents the desired configuration of registration controller installed on hub.
+type RegistrationComponent struct {
+	CommonComponentConfig `json:",inline" protobuf:"bytes,1,opt,name=commonComponentConfig"`
+}
+
+// CommonComponentConfig represents the common configuration of the component.
+type CommonComponentConfig struct {
+	// Replicas is the number of desired replicas
+	Replicas int32 `json:"replica,omitempty" protobuf:"varint,1,opt,name=replicas"`
```

Comment:
this looks questionable to me and as a concept will prevent a proper HPA from functioning.

Reply:
i am not sure we need hpa here, but I agree with you that by default we should have 3 replicas with leader elect.

Reply:
I think we can remove this for now and explore exact needs for this concern in the future, let's just use leader election with 3 replicas to start